### PR TITLE
[BEAM-9919] Added an External Transform API to Go SDK

### PR DESCRIPTION
**Pull request by pskevin**
It is now possible to execute an External (a.k.a, Cross-language) Transform within pipelines authored using the Go SDK. 
It also includes a Word count example run in LOOPBACK mode using the test python expansion service and executed by Python's portable runner.

---------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by pskevin** [comment]
R: @lostluck

**Comment by pskevin** [comment]
CC: @chamikaramj @robertwb @youngoli

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ValidateRunnerXlangTest.java [hunk 1]
```diff
@@ -177,6 +232,14 @@ public void coGroupByKeyTest() {
     PAssert.that(col).containsInAnyOrder("0:1,2,4", "1:3,5,6");
   }
 
+  /**
+   * Motivation behind combineGloballyTest.
```
Comment (chamikaramj) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ValidateRunnerXlangTest.java:236]: Let's keep these unrelated documentation updates in a separate PR.

File: sdks/go/examples/xlang/wordcount/input [hunk 2]
```diff
@@ -0,0 +1,5 @@
+Lorem ipsum dolor sit amet, consectetur adipiscing elit. Mauris id tellus vehicula, rutrum turpis quis, suscipit est. Quisque vehicula nec ex a interdum. Phasellus vulputate nunc sit amet nisl dapibus tincidunt ut ullamcorper nisi. Mauris gravida porta leo vel congue. Duis sit amet arcu eu nisl pharetra interdum a eget enim. Nulla facilisis massa ut egestas interdum. Nunc elit dui, hendrerit at pharetra a, pellentesque non turpis. Integer auctor vulputate congue. Pellentesque habitant morbi tristique senectus et netus et malesuada fames ac turpis egestas. Ut sagittis convallis lorem non semper. Ut ultrices elit a enim pulvinar fermentum.
```
Comment (chamikaramj) [inline sdks/go/examples/xlang/wordcount/input:1]: Is it possible to use a generated input for tests instead of committing this file ?

File: sdks/go/examples/xlang/wordcount/xlang_wordcount.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
+package main
```
Comment (chamikaramj) [inline sdks/go/examples/xlang/wordcount/xlang_wordcount.go:1]: You need to add Apache license to all new files.

File: sdks/go/examples/xlang/wordcount/xlang_wordcount.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	s := p.Root()
+
+	lines := textio.Read(s, *input)
+	// Convert lines of text into individual words.
+	col := beam.ParDo(s, extractFn, lines)
+
+	// Using Cross-language Count from Python's test expansion service
+	// TODO(pskevin): Cleaner using-face API
+	outputType := typex.NewKV(typex.New(reflectx.String), typex.New(reflectx.Int64))
+	external := &beam.ExternalTransform{
```
Comment (chamikaramj) [inline sdks/go/examples/xlang/wordcount/xlang_wordcount.go:83]: Woohoo!

File: sdks/go/examples/xlang/wordcount/xlang_wordcount.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+		In:            []beam.PCollection{col},
+		Urn:           "beam:transforms:xlang:count",
+		ExpansionAddr: "localhost:8118",
+		Out:           []typex.FullType{outputType},
+		Bounded:       true, // TODO(pskevin): Infer this value from output PCollection(s) part of the expanded tranform
+	}
+	counted := beam.CrossLanguage(s, p, external) // TODO(pskevin): Add external transform to Pipeline without passing it to the transform
```
Comment (chamikaramj) [inline sdks/go/examples/xlang/wordcount/xlang_wordcount.go:90]: Can we directly setup 'CrossLanguage' transform as the primary user API instead of having to setup a separate ExternalTransform struct ?

File: sdks/python/Pipfile [hunk 6]
```diff
@@ -0,0 +1,12 @@
+[[source]]
+name = "pypi"
```
Comment (chamikaramj) [inline sdks/python/Pipfile:2]: Do we need this ?

File: sdks/python/Pipfile.lock [hunk 7]
```diff
@@ -0,0 +1,946 @@
+{
+    "_meta": {
+        "hash": {
+            "sha256": "4d830f0b6c127288985ac4b35fef456c1bfcd8ffb2d771f32593d302e04bb80c"
+        },
+        "pipfile-spec": 6,
+        "requires": {
```
Comment (chamikaramj) [inline sdks/python/Pipfile.lock:7]: Ditto.

File: sdks/python/apache_beam/examples/wordcount_docker.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
```
Comment (chamikaramj) [inline sdks/python/apache_beam/examples/wordcount_docker.py:4]: Probably extract out these wordcount_xxx updates to a separate PR.

File: sdks/python/apache_beam/transforms/validate_runner_xlang_test.py [hunk 9]
```diff
@@ -14,6 +14,42 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 #
+
+"""
+###########################################################
+Runner Validation Test Suite for Cross-language Transforms
+###########################################################
+ As per Beams's Portability Framework design, Cross-language transforms
```
Comment (chamikaramj) [inline sdks/python/apache_beam/transforms/validate_runner_xlang_test.py:22]: Ditto regarding keeping these documentation updates in the other PR.

File: sdks/go/pkg/beam/external.go [hunk 10]
```diff
@@ -16,10 +16,151 @@
...
+	*/
+	// No information about the output types and bounded nature has been explicitly passed by the user
+	if len(e.Out) == 0 || cap(e.Out) == 0 {
+		// Infer output types from ExpansionResponse and update e.Out
+		if e.Out == nil {
+			// Use reverse schema encoding
+		} else {
+			// Use the coders list and map from coder id to internal FullType?
+		}
+	}
```
Comment (lostluck) [inline sdks/go/pkg/beam/external.go:151]: Since the next change will handle PCollection mappings by user provided string keys, we can probably get rid of this for now. Do add a TODO(lostluck) to remind me to plug in the schema coders here instead of what we're doing for the output types.

File: sdks/go/pkg/beam/pipeline.go [hunk 11]
```diff
@@ -60,7 +60,8 @@ func (s Scope) String() string {
 // Pipelines can safely be executed concurrently.
 type Pipeline struct {
 	// real is the deferred execution Graph as it is being constructed.
-	real *graph.Graph
+	real               *graph.Graph
+	ExpandedTransforms map[string]*ExternalTransform
```
Comment (lostluck) [inline sdks/go/pkg/beam/pipeline.go:64]: As discussed, we don't want to have to hang the expanded transforms on the pipeline here.

File: sdks/go/pkg/beam/runners/universal/universal.go [hunk 12]
```diff
@@ -82,6 +82,54 @@ func Execute(ctx context.Context, p *beam.Pipeline) error {
 		return errors.WithContextf(err, "generating model pipeline")
 	}
 
+	// Adding Expanded transforms to their counterparts in the Pipeline
+	for id, external := range p.ExpandedTransforms {
```
Comment (lostluck) [inline sdks/go/pkg/beam/runners/universal/universal.go:86]: Similarly, this code should be moved into graphx/translate.go (though TBH this is substantial enough that having things isolated in function in a graphx/xlang.go which are then called in translate.go would be a good move.)

File: sdks/go/pkg/beam/runners/universal/universal.go [hunk 13]
```diff
@@ -82,6 +82,54 @@ func Execute(ctx context.Context, p *beam.Pipeline) error {
...
+		for _, t := range pipeline.Components.Transforms {
+			for idx, i := range t.Inputs {
+				if i == existingInput {
+					t.Inputs[idx] = newInput
+				}
+			}
+		}
+
+		// Adding components of the Expanded Transforms to the current Pipeline
+		for k, v := range external.Components.Transforms {
```
Comment (lostluck) [inline sdks/go/pkg/beam/runners/universal/universal.go:110]: You should be able to clean this up with pipelinex.Update

https://github.com/apache/beam/blob/master/sdks/go/pkg/beam/core/runtime/pipelinex/replace.go#L32

Note that you'll need to delete the "go" entry from the environments first to do that safely for the loopback mode fix.

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ValidateRunnerXlangTest.java (hunk 1 above)
Comment (pskevin) [reply runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ValidateRunnerXlangTest.java]: Resolving since this was caused due to botched git workflows and have since been corrected.

File: sdks/go/examples/xlang/wordcount/input (hunk 2 above)
Comment (pskevin) [reply sdks/go/examples/xlang/wordcount/input:1]: Makes sense! Ack.

File: sdks/go/examples/xlang/wordcount/xlang_wordcount.go (hunk 3 above)
Comment (pskevin) [reply sdks/go/examples/xlang/wordcount/xlang_wordcount.go:1]: Ack.

File: sdks/python/Pipfile (hunk 6 above)
Comment (pskevin) [reply sdks/python/Pipfile]: Resolving since this was caused due to botched git workflows and have since been corrected.

File: sdks/python/Pipfile.lock (hunk 7 above)
Comment (pskevin) [reply sdks/python/Pipfile.lock]: Resolving since this was caused due to botched git workflows and have since been corrected.

File: sdks/python/apache_beam/transforms/validate_runner_xlang_test.py (hunk 9 above)
Comment (pskevin) [reply sdks/python/apache_beam/transforms/validate_runner_xlang_test.py]: Resolving since this was caused due to botched git workflows and have since been corrected.

File: sdks/python/apache_beam/examples/wordcount_docker.py (hunk 8 above)
Comment (pskevin) [reply sdks/python/apache_beam/examples/wordcount_docker.py]: Resolving since this was caused due to botched git workflows and have since been corrected.

File: sdks/go/examples/xlang/wordcount/xlang_wordcount.go (hunk 5 above)
Comment (pskevin) [reply sdks/go/examples/xlang/wordcount/xlang_wordcount.go:90]: Ack.

All references to the previous API are just placeholders where (in the future) code to port the legacy API will go. They don't influence how the current API is being developed. In fact it's just a bonus that the previous API can be cleanly supported.

File: sdks/go/examples/xlang/wordcount/xlang_wordcount.go [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+		empty.Inc(ctx, 1)
+	}
+	for _, word := range wordRE.FindAllString(line, -1) {
+		emit(word)
+	}
+}
+
+// formatFn is a DoFn that formats a word and its count as a string.
+func formatFn(w string, c int64) string {
+	fmt.Println(w, c)
```
[the remark this answers is not in the document]
Comment (pskevin) [reply sdks/go/examples/xlang/wordcount/xlang_wordcount.go:51]: Thanks for catching that! Ack.

File: sdks/go/examples/xlang/wordcount/xlang_wordcount.go [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+	return fmt.Sprintf("%s: %v", w, c)
+}
+
+func init() {
+	beam.RegisterFunction(extractFn)
+	beam.RegisterFunction(formatFn)
+}
+
+func main() {
+	// If beamx or Go flags are used, flags must be parsed first.
```
[the remark this answers is not in the document]
Comment (pskevin) [reply sdks/go/examples/xlang/wordcount/xlang_wordcount.go:61]: Thanks for catching that! Will do.

File: sdks/go/examples/xlang/wordcount/xlang_wordcount.go [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+		ExpansionAddr: "localhost:8118",
+		Out:           []typex.FullType{outputType},
+		Bounded:       true, // TODO(pskevin): Infer this value from output PCollection(s) part of the expanded tranform
+	}
+	counted := beam.CrossLanguage(s, p, external) // TODO(pskevin): Add external transform to Pipeline without passing it to the transform
+
+	formatted := beam.ParDo(s, formatFn, counted[0])
+	textio.Write(s, *output, formatted)
+
+	// Concept #1: The beamx.Run convenience wrapper allows a number of
```
[the remark this answers is not in the document]
Comment (pskevin) [reply sdks/go/examples/xlang/wordcount/xlang_wordcount.go:95]: Ack.

File: sdks/go/pkg/beam/external.go [hunk 17]
```diff
@@ -16,10 +16,151 @@
...
+
+// CrossLanguage is the temporary API to execute external transforms
+// TODO(pskevin): Handle errors using the TryN and Must strategies instead one function handling multiple points of failure
+func CrossLanguage(s Scope, p *Pipeline, e *ExternalTransform) []PCollection {
+	if e.ExpansionAddr == "" { // TODO(pskevin): Better way to check if the value was ever set
+		// return Legacy External API
+	}
+
+	/*
+		Add ExternalTranform to the Graph
```
[the remark this answers is not in the document]
Comment (pskevin) [reply sdks/go/pkg/beam/external.go:52]: I'm definitely missing something since I can't see any typo. Could you be more explicit?

Also, the block comments were meant just to separate sections of independent logic which I expect to break down into functions during clean up. Should I remove all the block comments?

File: sdks/go/pkg/beam/external.go [hunk 18]
```diff
@@ -16,10 +16,151 @@
...
+	*/
+	// Validating scope and inputs
+	if !s.IsValid() {
+		// return nil, errors.New("invalid scope")
```
[the remark this answers is not in the document]
Comment (pskevin) [reply sdks/go/pkg/beam/external.go:56]: Ack.

File: sdks/go/pkg/beam/external.go [hunk 19]
```diff
@@ -16,10 +16,151 @@
...
+		Namespace:  s.String(),
+	}
+
+	/*
+		Querying Expansion Service
+	*/
+	// Setting grpc client
+	conn, err := grpc.Dial(e.ExpansionAddr, grpc.WithInsecure())
+	if err != nil {
+		panic(err)
```
[the remark this answers is not in the document]
Comment (pskevin) [reply sdks/go/pkg/beam/external.go:126]: Ack! Thanks for also exemplifying the correct function to use.

File: sdks/go/pkg/beam/external.go (hunk 10 above)
Comment (pskevin) [reply sdks/go/pkg/beam/external.go:151]: Makes sense. Will do!

File: sdks/go/pkg/beam/runners/universal/universal.go (hunk 13 above)
Comment (pskevin) [reply sdks/go/pkg/beam/runners/universal/universal.go:110]: That's an awesome helper. Thanks for pointing it out.

File: sdks/go/pkg/beam/external.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/external.go:52]: ExternalTranform should be ExternalTransform.

I'd prefer no block comments. They're uncommon in Go code.  

To be clear, commenting on what the next sections of code is what comments are for. They don't need to take up so much space though.

[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/external.go:52]: To be pedantically clear: No block comments doesn't mean no comments. In my original comment, I suggested the line comment alternative.

File: sdks/go/examples/xlang/wordcount/xlang_wordcount.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pskevin) [reply sdks/go/examples/xlang/wordcount/xlang_wordcount.go:17]: Ack.

File: sdks/go/pkg/beam/external.go [hunk 20]
```diff
@@ -16,10 +16,151 @@
...
+	Bounded           bool
+	ExpansionAddr     string
+	Components        *pipepb.Components
+	ExpandedTransform *pipepb.PTransform
+	Requirements      []string
+}
+
+// CrossLanguage is the temporary API to execute external transforms
+// TODO(pskevin): Handle errors using the TryN and Must strategies instead one function handling multiple points of failure
+func CrossLanguage(s Scope, p *Pipeline, e *ExternalTransform) []PCollection {
```
[the remark this answers is not in the document]
Comment (pskevin) [reply sdks/go/pkg/beam/external.go:45]: Ack.

File: sdks/go/pkg/beam/external.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (pskevin) [reply sdks/go/pkg/beam/external.go]: Thanks for clarifying in detail. Reflected the changes.

**Comment by pskevin** [comment]
I've addressed the initial comments with the exception of ones suggesting organizational changes. In general this discussion will dictate where the final (and cleaner) API will reside and how the user will make use of it. Since I won't be the stakeholder for this feature after a few weeks, I thought it'd be worthwhile bringing in more opinions – @lostluck @robertwb @youngoli 

The challenges and possible solutions are as follows:

- Where should all xlang code be?\
There are multiple moving parts related to the execution of cross-language transforms which are now condensed into one `TryCrossLanguage` function in `external.go`  

	- Allowing various combinations of inputs/outputs (n:n, 1:0, 0:1)
	- Correctly adding `ExternalTransform` to `Graph` representing the unexpanded cross-language transform within the pipeline
	- Building the `ExpansionRequest`
	- Querying the expansion service for the corresponding `ExpansionResponse`
	- Querying the expansion service for Artifact Resolution and Staging
	- Handling the expanded components of the `ExternalTransform`

	Decomposing these functions correctly within package(s) dedicated to `ExternalTransform` whether as `beam/.../xlangx` or `beam/.../xlang` or both will be ideal. It helps in segregation of responsibilities so that each function can be developed to maturity independent of each other. I can then make incremental progress on different fronts without compromising overall functionality that this PR has already achieved.  
	Futhermore, `external.go` can then use the package(s) to expose generic APIs such as `beam.CrossLanguage` (for n:n inputs/outputs), `beam.CrossLanguageAsSource` (for 0:1 inputs/outputs) and `beam.CrossLanguageAsSink` (for 1:0 inputs/outputs).  
	In the future, when the existing API is to be deprecated, these functions could then be renamed to their `beam.ExternalTransform`... counterparts. The exisiting API could be renamed to `beam.LegacyExternalTransform` and implemented using the API that exists as the outcome of this PR (since it is already backwards compatible for the most part).  

	@lostluck already pointed to a possible segregation by removing pipeline handling code within `universal.go` into `graphx/xlang.go` to be used in `graphx/translate.go`. I could explore similar options in alignment to the results of this discussion.

- How is `ExternalTransform` associated to pipeline construction?
	- Adding to `Graph` and/or `MultiEdge`   
	Currently, `graph.Opcode` for external transforms already exists (with value as `"External"`) and is used in conjunction with `graph.Payload` by the existing external transform API. Adding `ExternalTransform` to `Graph` will pull in proto and grpc dependencies within `beam/core/graph` regardless of whether the pipeline contains any cross-language transform or not. Adding `ExternalTransform` to `MultiEdge` will need additional logic to distinguish between an external transform using the proposed v/s current API.  
	One scanario in which  I believe this method could be useful is if correct pipeline construction required explicit handling of `ExternalTransform` when each `MultiEdge` is being added during the `graphx.Marshal` call. As of now I don't feel that is the case.

	- Augmenting `Pipeline` and `Scope`  
	As is implemented in this PR, `ExpandedTransforms map[string]ExternalTransform` field is added to `Pipeline`. The key is the `MultiEdge`'s ID that represents an external transform within `Graph`. Using this, during pipeline construction, the Go proto representing this transform is swapped with the expanded `ExternalTransform`. All of the components associated with the expanded `ExternalTransform` are also added to the pipeline proto. This method does not conflict with any pipeline representation/construction code for strictly Go pipelines and may be more suitable for the future.  

		An existing problem worth calling out is the necessity to pass in the `Pipeline` reference to `beam.CrossLanguage` to add the new `ExternalTransform` in the `ExpandedTransforms` map. A possible solution to this would be:
		1. Adding `func (p *Pipeline) RegisterExternalTransform(k string, e *ExternalTransform)` function to `Pipeline`
		2. Adding `AddMapEntry func(string, *ExternalTransform)` field to `Scope`
		3. Changing `func (p *Pipeline) Root() Scope` to ` return Scope{scope: p.real.Root(), real: p.real, AddMapEntry: p.RegisterExternalTransform}`  

		`beam.CrossLanguage` should then be able to use `s.AddMapEntry(key, externalTransform)` instead of the current implementation.

File: sdks/go/pkg/beam/external.go [hunk 21]
```diff
@@ -16,10 +16,151 @@
...
+	if e.ExpansionAddr == "" { // TODO(pskevin): Better way to check if the value was ever set
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/external.go:50]: Note, by this comment, the intent was for you to remove this dead code, as it's unnecessary.

File: sdks/go/pkg/beam/external.go [hunk 22]
```diff
@@ -16,10 +16,144 @@
...
 	"github.com/apache/beam/sdks/go/pkg/beam/core/graph"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/runtime/graphx"
 	"github.com/apache/beam/sdks/go/pkg/beam/internal/errors"
+	jobpb "github.com/apache/beam/sdks/go/pkg/beam/model/jobmanagement_v1"
+	pipepb "github.com/apache/beam/sdks/go/pkg/beam/model/pipeline_v1"
+	"google.golang.org/grpc"
 )
 
+// ExternalTransform represents the cross-language transform in and out of the Pipeline as a MultiEdge and Expanded proto respectively
+type ExternalTransform struct {
```
Comment (lostluck) [inline sdks/go/pkg/beam/external.go:31]: Per the other thread, please move these to parameters on CrossLanguage and TryCrossLanguage instead. Do not try to force in compatibility with the legacy External, it's OK for them to have two separate calls and paths.
By having them as a struct it's not clear what is required and what is not, and the compiler won't help the user by failing at compile time. 

An aside: The other issue here is you've mixed up user side parameters with internal implementation details, and made them part of the API surface. APIs are easiest to use when the user knows how to fill everything and what is required or not. The components and Expanded transform and requirements fields are not something that users would be filling in for example. Types are cheap. Make a new type instead of trying to reuse something that almost fits.

File: sdks/go/pkg/beam/external.go [hunk 23]
```diff
@@ -16,10 +16,144 @@
...
+	id                int
+	Urn               string
+	Payload           []byte
+	In                []PCollection
+	Out               []FullType
+	Bounded           bool
+	ExpansionAddr     string
+	Components        *pipepb.Components
+	ExpandedTransform *pipepb.PTransform
+	Requirements      []string
```
Comment (lostluck) [inline sdks/go/pkg/beam/external.go:41]: Move these to a graph.CrossLanguage struct, but have their proto types be interface{} instead, with a comment about what the types should be. Given those fields are only used by beam framework internals, there's little risk in using type assertions for them in the right places, such as the graphx package.

File: sdks/go/pkg/beam/external.go [hunk 24]
```diff
@@ -16,10 +16,151 @@
...
+	}
+	defer conn.Close()
+	client := jobpb.NewExpansionServiceClient(conn)
+
+	// Handling ExpansionResponse
+	res, err := client.Expand(context.Background(), req)
+	if err != nil {
+		panic(err)
+	}
+	e.Components = res.GetComponents()
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/pkg/beam/external.go:133]: I don't know why it didn't occur to me before, but if the deserialized proto object needs to be "stored" on an object in the graph package, that field can just be an interface{} and be commented that it's expected to be the proto type. As the fields are only usable by beam framework internals, it's fine to assume that they'll be the correct types by construction (validated by unit tests at some juncture).

File: sdks/go/pkg/beam/external.go (hunk 22 above)
Comment (pskevin) [reply sdks/go/pkg/beam/external.go:31]: The current version was never meant to be representative of what the final API would be. Only after discussing code organization was the API surface meant to be refined. 

Thanks for the new type idea! It'll make separation of concerns much clearer.

File: sdks/go/pkg/beam/external.go (hunk 23 above)
Comment (pskevin) [reply sdks/go/pkg/beam/external.go:41]: Awesome suggestion. Will work on it.

File: sdks/go/examples/xlang/wordcount/xlang_wordcount.go [hunk 25]
```diff
@@ -0,0 +1,107 @@
...
+	lines := textio.Read(s, *input)
+	col := beam.ParDo(s, extractFn, lines)
+
+	// Using Cross-language Count from Python's test expansion service
+	// TODO(pskevin): Cleaner using-face API
+	outputType := typex.NewKV(typex.New(reflectx.String), typex.New(reflectx.Int64))
+	external := &beam.ExternalTransform{
+		In:            []beam.PCollection{col},
+		Urn:           "beam:transforms:xlang:count",
+		ExpansionAddr: "localhost:8118",
```
[the remark this answers is not in the document]
Comment (lostluck) [reply sdks/go/examples/xlang/wordcount/xlang_wordcount.go:95]: Alas, it's got to be a parameter. Users may request transforms from multiple independent expansion servers (eg. a Java one, a python one, a dataflow one...)

[the remark this answers is not in the document]
Comment (youngoli) [reply sdks/go/examples/xlang/wordcount/xlang_wordcount.go:95]: Gotcha, but it might still be a good idea in this example if the expansion servers' addresses can change. I'd expect that if you were a user writing and running cross language pipelines, you wouldn't want to change the address in your code each time, you'd want some flag you can change instead. Same if you have multiple different expansion servers.

Of course, if expansion servers are expected to have a consistent address then this seems good as-is.

[the remark this answers is not in the document]
Comment (pskevin) [reply sdks/go/examples/xlang/wordcount/xlang_wordcount.go]: Thanks for that suggestion @youngoli! I've updated it.

File: sdks/go/pkg/beam/external.go [hunk 26]
```diff
@@ -16,10 +16,151 @@
...
+		ins = append(ins, col.n)
+	}
+	edge := graph.NewCrossLanguage(s.real, s.scope, ins, payload)
+
+	// TODO(pskevin): There needs to be a better way of associating this ExternalTransform to the pipeline
+	// Adding ExternalTransform to pipeline referenced by MultiEdge ID
+	if p.ExpandedTransforms == nil {
+		p.ExpandedTransforms = make(map[string]*ExternalTransform)
+	}
+	p.ExpandedTransforms[fmt.Sprintf("e%v", edge.ID())] = e
```
[the remark this answers is not in the document]
Comment (pskevin) [reply sdks/go/pkg/beam/external.go:76]: Ack. Updated!

File: sdks/go/pkg/beam/external.go (hunk 24 above)
[the remark this answers is not in the document]
Comment (pskevin) [reply sdks/go/pkg/beam/external.go:127]: That's a great suggestion. I've updated it to be in line with it.

File: sdks/go/pkg/beam/pipeline.go (hunk 11 above)
Comment (pskevin) [reply sdks/go/pkg/beam/pipeline.go]: Removed from pipeline.

File: sdks/go/pkg/beam/runners/universal/universal.go (hunk 12 above)
Comment (pskevin) [reply sdks/go/pkg/beam/runners/universal/universal.go]: I've reorganized much of the code. Your opinions would be helpful to understand which of them are not relevant.

File: sdks/go/pkg/beam/core/graph/xlang.go [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+
+func (ext ExternalTransform) WithNamedInputs(inputs map[string]*Node) ExternalTransform {
+	if ext.inputs != nil {
+		panic(errors.Errorf("inputs already set as: \n%v", ext.inputs))
+	}
+	ext.inputs = inputs
+	return ext
+}
+
+func (ext ExternalTransform) WithNamedOutputs(outputTypes map[string]typex.FullType) ExternalTransform {
```
Comment (lostluck) [inline sdks/go/pkg/beam/core/graph/xlang.go:82]: Good use of a value method here and the similar functions.

**Comment by pskevin** [comment]
@lostluck My updates are ready to review!

**Comment by pskevin** [comment]
Thanks a ton @lostluck!
